filters/enforcescale.c

```diff
@@ -7,40 +7,69 @@ MFD_FILTER(enforcescale)
 			lv2:scalePoint [ rdfs:label "Always up"  ; rdf:value 2 ] ;
 			rdfs:comment "Behaviour towards of off-key notes."
 			)
+	, TTF_IPORT(1, "scale", "Scale",  0, 11, 0,
+			lv2:portProperty lv2:integer; lv2:portProperty lv2:enumeration;
+			lv2:scalePoint [ rdfs:label "C"  ; rdf:value 0 ] ;
+			lv2:scalePoint [ rdfs:label "C#" ; rdf:value 1 ] ;
+			lv2:scalePoint [ rdfs:label "D"  ; rdf:value 2 ] ;
+			lv2:scalePoint [ rdfs:label "D#" ; rdf:value 3 ] ;
+			lv2:scalePoint [ rdfs:label "E"  ; rdf:value 4 ] ;
+			lv2:scalePoint [ rdfs:label "F"  ; rdf:value 5 ] ;
+			lv2:scalePoint [ rdfs:label "F#" ; rdf:value 6 ] ;
+			lv2:scalePoint [ rdfs:label "G"  ; rdf:value 7 ] ;
+			lv2:scalePoint [ rdfs:label "G#" ; rdf:value 8 ] ;
+			lv2:scalePoint [ rdfs:label "A"  ; rdf:value 9 ] ;
+			lv2:scalePoint [ rdfs:label "A#" ; rdf:value 10 ] ;
+			lv2:scalePoint [ rdfs:label "B"  ; rdf:value 11 ] ;
+			rdfs:comment "Limit note-on/off messages to this scale."
+			)
+    , TTF_IPORT(3, "scaleMode", "ScaleMode",  0, 6, 0, \
+			lv2:portProperty lv2:integer; lv2:portProperty lv2:enumeration;
+			lv2:scalePoint [ rdfs:label "Ionian/Major"  ; rdf:value 0 ] ;
+			lv2:scalePoint [ rdfs:label "Aeolian/Minor" ; rdf:value 1 ] ;
+			lv2:scalePoint [ rdfs:label "Dorian"  ; rdf:value 2 ] ;
+			lv2:scalePoint [ rdfs:label "Phrygian"  ; rdf:value 3 ] ;
+			lv2:scalePoint [ rdfs:label "Lydian" ; rdf:value 4 ] ;
+			lv2:scalePoint [ rdfs:label "Mixolydian"  ; rdf:value 5 ] ;
+			lv2:scalePoint [ rdfs:label "Locrian"  ; rdf:value 6 ] ;
+			rdfs:comment "Mode of the Scale"
+			)
 	; rdfs:comment "Filter note-on/off events depending on musical scale. If the key is changed note-off events of are sent for all active off-key notes." ; 
 	.
 
 #elif defined MX_CODE
+/*
+ * Mode Tonic relative
+ * scale Interval sequence Example
+ * Ionian     I     W–W–H–W–W–W–H     C–D–E–F–G–A–B–C
+ * Dorian     II    W–H–W–W–W–H–W     D–E–F–G–A–B–C–D
+ * Phrygian   III   H–W–W–W–H–W–W     E–F–G–A–B–C–D–E
+ * Lydian     IV    W–W–W–H–W–W–H     F–G–A–B–C–D–E–F
+ * Mixolydian V     W–W–H–W–W–H–W     G–A–B–C–D–E–F–G
+ * Aeolian    VI    W–H–W–W–H–W–W     A–B–C–D–E–F–G–A
+ * Locrian    VII   H–W–W–H–W–W–W     B–C–D–E–F–G–A–B
+ *
+ */
+
+static int filter_enforcescale_check(int scale,int scalemode, uint8_t key) {
+	const short scales[7][12] = {
+		{ 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1 },  // Ionian/Major
+		{ 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 0 },  // Aeolian/Minor
+		{ 1, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0 },  // Dorian
+		{ 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0 },  // Phrygian
+		{ 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1 },  // Lydian
+		{ 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0 },  // Mixolydian
+		{ 1, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0 },  // Locrian
 
-
-static int filter_enforcescale_check(int scale, uint8_t key) {
-	const short major_scale[12] = {
-		1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 0, 1
 	};
-	return major_scale[(key - scale + 12) % 12];
+	return scales[scalemode][(key - scale + 12) % 12];
```

Comment:
You could simply have used an offset here using the `major_scale` and modulo operation only.

Reply:
Ok , i see , two independent  input ports , double the hassle ..... i wasn't automating scalechanges and focused on the ui aspect of easily setting desired root/mode combinations. Also the major_scale & offset method can make all this more implicit. Have also looked in the chord.c filter , that has also only Major scales selectable in the UI . Therein you explicitly  enumerated all intervall_halfnote_distances for each of the Major modes. Maybe there could also work some offset-computational-aproach .

Reply:
I was thinking of  `return major_scale[(key - scale + mode + 12) % 12];`

Where `mode` is 2 for Dorian, 4 for Phrygian, 5 for Lydian, etc. The port value would be 0..11 but only 7 scale-points be defined.

Anyway. I still don't like to have two ports that effectively control the same parameter.

Reply:
All right, got it. Bur with respect to "exotic scales" which can't be easily accesed through shifting the major scale pattern, like pentatonic , octatonic , neapolian , ..... it would make sense to keep the "scales matrix" ... to incorporate all scales which exist e.g. in the pianoroll of qtractor, we need a more general aproach to express these. I see you only want to have one input port , maybe i should keep enforcescale as is, to keep backwards compatibility , and create a new filter module "scaleizer" with access to more "exotic scales/modes/keysets" . There one has to also fiddle with the transp value. I want to enrich this in your sense, so tell me waht you think about this.

Reply:
Last I looked, qtractor used http://www.huygens-fokker.org/scala/ and I have no intention to add support for this in a  plugin. This is something for hosts to provide, ideally also to plugins. 

This current "Enforce Scale" is just to make a very common case fast.  You can have a single control-knob, useful when playing live.

For a flexible plugin that can handle any 12tone scale, the idea there would be have 12 toggle inputs that allow to select allowed notes, perhaps combined with a chromatic transpose offset. 

It could also have alternative option that allows to learn allowed notes from a chord instead of control input ports:

e.g. CC64 pedal down; play a chord or cluster of allowed notes; pedal-up.  That would be useful in a live context as well.

What do you think?

A custom GUI, like the MOD, could have a some UI abstraction that lists well known scales. It could also just be presets! There's ready-made support for those already and LV2 presets can be partial and only contain the scale. The root-note can be kept separately.

Anyway, I'd rather define scales in the .ttl (a user can extend them) instead of hardcoding them in the source.